### PR TITLE
Upgrade Keycloak and add tests

### PR DESCRIPTION
**Pull request by ajoel24**
Changelist:
- Minor code refactor
- Upgrade Keycloak version from `16.1.1` to `18.0.0`
- Add a simple test to ensure hashing is working as expected

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -3,13 +3,20 @@ plugins {
...
 }
 
+configurations {
+    testImplementation.get().apply {
+        extendsFrom(configurations.compileOnly.get())
+    }
+}
+
 dependencies {
+    testImplementation("org.junit.jupiter:junit-jupiter-api:5.8.2")
```
Comment (leroyguillaume) [inline build.gradle.kts:19]: Can you move this line after the `Keycloak` block please?

File: src/test/java/com/github/leroyguillaume/keycloak/bcrypt/BCryptPasswordHashProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    @DisplayName("Should hash the password successfully")
+    void shouldHashThePasswordSuccessfully() {
+        String rawPassword = "test";
+        String hashedPassword = provider.encode(rawPassword, iterations);
+        PasswordCredentialModel model = PasswordCredentialModel.createFromValues(id, new byte[0], iterations, hashedPassword);
+
+        assertNotNull(hashedPassword);
+        assertTrue(provider.verify(rawPassword, model));
+    }
+}
```
Comment (leroyguillaume) [inline src/test/java/com/github/leroyguillaume/keycloak/bcrypt/BCryptPasswordHashProviderTest.java:24]: Can you add an empty line at the end of the file please?

**Review by leroyguillaume — CHANGES_REQUESTED** [review]
Thanks for your contribution! Need few changes.

**Comment by ajoel24** [comment]
Sure @leroyguillaume . Will do the changes.
I noticed when I used  this provider with Keycloak v18.0.x, I am unable to create users through the Keycloak UI. It is working for Keycloak REST API.


To be elaborate, in Keycloak UI, once we set the password policy as bcrypt, any new users created will not work since the password is not being stored properly. I checked the database table `credentials` where the password is not being stored. However, this is not the case for REST API.

Can you please check if this issue exists in this version of Keycloak?
If so, we might have to check the issues page in Keycloak and raise one if not exists.
Thanks.

File: build.gradle.kts (hunk 1 above)
Comment (ajoel24) [reply build.gradle.kts]: Done.

File: build.gradle.kts [hunk 3]
```diff
@@ -45,4 +52,8 @@ tasks {
     wrapper {
         gradleVersion = "6.4"
     }
-}
+
+    test {
+        useJUnitPlatform()
+    }
+}
```
[the remark this answers is not in the document]
Comment (ajoel24) [reply build.gradle.kts]: Done

File: docker/Dockerfile [hunk 4]
```diff
@@ -2,6 +2,6 @@ ARG keycloak_version
 
 FROM quay.io/keycloak/keycloak:${keycloak_version}
 
-ARG keycloak_bcrypt_version=1.5.0
+ARG keycloak_bcrypt_version=1.5.1
```
[the remark this answers is not in the document]
Comment (ajoel24) [reply docker/Dockerfile:5]: Done

File: gradle.properties [hunk 5]
```diff
@@ -1 +1 @@
-dependency.keycloak.version=16.1.1
+dependency.keycloak.version=18.0.0
```
[the remark this answers is not in the document]
Comment (ajoel24) [reply gradle.properties]: Done

File: src/test/java/com/github/leroyguillaume/keycloak/bcrypt/BCryptPasswordHashProviderTest.java (hunk 2 above)
Comment (ajoel24) [reply src/test/java/com/github/leroyguillaume/keycloak/bcrypt/BCryptPasswordHashProviderTest.java]: Done

**Comment by ajoel24** [comment]
I also observed that Keycloak stores the hashes in `credentials` table with Base64 encoding. So, we need to include Base64 encoding / decoding in our implementation as well.

**Comment by leroyguillaume** [comment]
@ajoel24 I tried to reproduce you bug:
- I started KC v16.1.1 with postgres database
- I configured the password policy with BCrypt
- I created `test` user with `changeit` password
- I stopped KC
- I upgraded to KC v18.0.0 (with the same postgres database of course)
- I tested to login with `test`/`changeit` credentials and it is OK
- I created a new user `test2` with `changeit` password
- I tested to login with `test2`/`changeit` credentials and it is OK

In the database, I've:
```
keycloak=# select * from credential;
-[ RECORD 1 ]---+-------------------------------------------------------------------------------------------------------------
id              | 96e06dbf-9898-45c5-b97d-8a476f2ae9be
salt            | 
type            | password
user_id         | ed7fc3e4-cd1f-4fef-af17-02687fd4db9d
created_date    | 1658655961164
user_label      | 
secret_data     | {"value":"$2y$10$yqK/laAraeNK7WROctsoWe39vADNEcyoH8KSaGwVRi2stUTQQ31YO","salt":"","additionalParameters":{}}
credential_data | {"hashIterations":-1,"algorithm":"bcrypt","additionalParameters":{}}
priority        | 10
-[ RECORD 2 ]---+-------------------------------------------------------------------------------------------------------------
id              | 74b381bb-0abb-4a0d-82cd-bba01a48633c
salt            | 
type            | password
user_id         | d0fff0ce-7ae6-46ba-aac2-6f1dcc83081e
created_date    | 1658655902759
user_label      | 
secret_data     | {"value":"$2y$10$WbOzPlMWhhWIxvSMZU4.YehSGmkKX76DBgjlRNTWHE6s5rhFLplkC","salt":"","additionalParameters":{}}
credential_data | {"hashIterations":-1,"algorithm":"bcrypt","additionalParameters":{}}
priority        | 10
-[ RECORD 3 ]---+-------------------------------------------------------------------------------------------------------------
id              | 39c742b9-278c-46a6-91c5-e8367546aec5
salt            | 
type            | password
user_id         | efb38b96-5ccf-4907-a82b-6f84e30f6c92
created_date    | 1658658268439
user_label      | 
secret_data     | {"value":"$2y$10$kdPqRFX03M5WS9.OEOP9vusWxQsiHrRt.kOQzKB6Ahnurfo0V2pea","salt":"","additionalParameters":{}}
credential_data | {"hashIterations":-1,"algorithm":"bcrypt","additionalParameters":{}}
priority        | 10
```

So, unfortunately, I think there's a problem on your side... 

BTW, I pushed on master an improvement of docker-compose to use KC v18+ in prod mode with postgresql to make the test more iso-prod

**Review by leroyguillaume — APPROVED** [review]

**Comment by leroyguillaume** [comment]
Oh I finally found the problem: it is in your PR. I must revert the change. 

The BCrypt base64 hash is not standard and include some characters (https://security.stackexchange.com/questions/22627/base64-encode-differences-in-bcrypt-implemenations).

So the line causes the bug:
``java
Base64.getEncoder().encodeToString(hash)
```

**Comment by ajoel24** [comment]
@leroyguillaume Thanks for the update.
I see this PR has been reverted through PR #29 .
Can you please suggest the changes that are required from my end?
Thanks.

**Comment by leroyguillaume** [comment]
@ajoel24 I closed the PR but I don't merge it, I just revert the Base64 part: https://github.com/leroyguillaume/keycloak-bcrypt/commit/00f2e36b84d6c37c254eb7bdb1445270b625a76b

**Comment by ajoel24** [comment]
@leroyguillaume Understood. Thanks.

**Pull request merged**